### DISCUSSION TITLE
Item Presentation View - A new extension type

### DISCUSSION THREAD

**Discussion by danieljwestman** [directus/directus#4930] [Feature Requests]
> _Maybe there is a similar discussion/issue already, let me know._

I think it would be nice to have a "Item Presentation" view in the App. A view where you _can_ provide a customized presentation (read only) of an collection item.

**Today** the hierarchy is like:  `Module` -> `Layout` -> `Edit item` (`Interfaces`)

**My idea** is: `Module` -> `Layout` -> `Item Presentation` -> `Edit item` (`Interfaces`)

- The presentation view should be optional - an option when created/updating a collection (like the Display option for a field).
- There could be a default/basic presentation which only provide a view with all the item `Displays`.
- You can create your own extensions for custom `Presentations`, just like `Modules`, `Layouts`, `Displays` and `Interfaces`.
- In the presentation view there should be some kind of "Edit button" which routes to the edit view (the interfaces). 
- Maybe you could use the `Displays` in the `Presentation` in some way

With this new concept you can easily build a more full-fledged consumer web app, not "only" a data manager with item edit views.

What do you think?

_(Obviously you could build a whole new Module to accomplish this, but that's more complicated and you will then often need to reinvent a lot of "core" stuff)_

**Comment by benhaynes** [comment]
Interesting! The main thing here is keeping the code, interfaces, concept, and setup as clean/simple as possible. We don't want these new ideas to make the core usage of the App more complicated. A few thoughts on this:

* **Displays** are meant to be inline, so they work best in small spaces... like a cell of the table Layout, or a color dot in the page title. I'm not sure that Displays are a good way to present _rich_ data on the detail page.
* **Interfaces** would be the preferred way to accomplish this as of now. Using permissions or the field's read-only property, you can set fields to be a presentation mode. 
    * _Conditional Interface UI_ — One option here would be to give our interfaces an alternate UI when in read-only mode. In this way, the interface could be/look completely different when read-only... instead of simply being a disabled version.
    * _Dedicated Presentation Interfaces_ — The other option would be to create a few key Presentation Interfaces that are only meant to _show_ data. So a big placard for dynamic numbers, or a spark chart, or a badge, etc. The only issue here is that they wouldn't have an edit mode (other than the built-in "Raw Value" field option. However, we _could_ consider adding a field option that allows you to choose a **different** interface for read-only mode. Then you could have a text-input field... with and big placard interface override in read-only mode.
* **Collections** could have an option for adding the "Presentation" stage to this flow. When enabled, navigating to an Item Detail page would show all fields as read-only... and add an "Edit" button to the header (for all fields), and/or to each Field (for editing individual fields). This seems like a simple, but powerful way to get a Presentation mode without having to restrict a Role's UPDATE permissions.

**Reply by danieljwestman** [reply]
Nice input. 

I still think separate views for presentation and editing would be an big improvement for building apps that's not only for admins. 

Dedicated, read-only, presentation interfaces could be a way, but my idea of the presentation of an item is that it should/could look pretty much different than the editing view, and in some cases not include all fields (as the interfaces when editing).

🤔

**Reply by Kematia** [reply]
What about Alias interfaces? It would be a great way to handle editor/public read-only content. 

Last week I had this use case:

Client needed a complex form on the frontend which has to be updated frequently. I set up a directus collection for each form, and basically replicated 1-to-1 the direcuts fields in the frontend using vuetify. I basically was using the directus collection builder as a WYSIWYG for the fronted, down to the field width, placeholders and read-only state! Which I think is pretty nifty but I digress.

One Issue was that we needed some read-only fields (for form section titles, notes, warnings etc), and I filled the gap with some Alias interfaces. This got me thinking that we could have interfaces that take either static or dynamic data, and "link" the item properties to pre-specified fields. 

Some examples:

- Title Alias (we have someting similar with the divider alias interface)
- Card Alias (similar to the card layout)
- Note Alias (a markdown/wysiwyg component to display more rich text)
- Custom Aliases (computed fields!)

We already have something VERY similar in the card layout... where you take a collection and specify what property is the image, which one is the title, subitlte etc... but instead of specifying it in a layout, you specify it in the collection directly.

A collection level "alias/preview" could be useful in m2x forms as well, where you could either use a the template-preview we have already or a richer preview, specified in the "alias template" section of a collection. This would pave the way to mixed collection layouts? The m2a builder would for example fetch either a card or a template, using whichever is specified in the collection. 

Another use case would be to show previews on mousehover when hovering a layout item. 

This is something I could see myself using a lot, both in the Admin App and in my frontend.


Hopefully I'm making sense!

**Reply by Kematia** [reply]
Another, more complex use case:

My client needed a "content" builder for a restaurant menu. It would have to display either a title, a note or a dish (fetched from a "dishes" collection. The m2a made the most sense for me, but the client had trouble navigating the interface: a title and a dish looked very similar in the layout. 

As I didn't care much for relationality, I spun up a custom interface that stored everything in a JSON blob, where I could assign a custom component to render the "dish" item in the list. The text would be displayed as a simple text-input, but the dish had a custom component that displayed it in a card style.

https://user-images.githubusercontent.com/17851386/114392852-a118f100-9b99-11eb-8473-3ac44646170b.mov

Now, if I had access to such a system, I would make every item editable in a drawer (akin to the m2a we have today) using the default directus fields, and use either the preview template for text, or the alias/display interface specified in the "dish" collection.
 
The custom component would have to only recieve the configuration, without having to be opinionated on which components to use. 

So with just creating custom "alias" interfaces, I could use the same custom component to build web pages, forms, marketing emails, etcetera, while giving my clients and editors a much easier time handling the content.

**Reply by benhaynes** [reply]
Nice custom interface! Would love your feedback on what was good/bad about the custom extension dev process.

**Reply by Kematia** [reply]
Thanks! Glad you like it :)
 I'm cleaning it up a bit then I'll open-source it, it's quite flexible! It's kind of an advanced repeater where you can specify what interfaces to use to build the repeater "row", even custom ones.

If you ever worked with prismic it's similar to theirs "dynamic zones", where in a item you have a set number of static fields, and then a zone where you have _n_ components you can add freely.

> Would love your feedback on what was good/bad about the custom extension dev process.

I had to pull in the directus app source code for some stuff (rendering the micromoustache for example), and I had to add some event listener on the drawer-collection... I was meaning to open a couple issues next week!

Not everything I needed was available in the NPM package.

**Reply by fa-briz-io** [reply]
Ciao @Kematia , your video preview looks very impressive. Hope you don't mind me asking, but just curious if you ever released the custom component? I'm trying to learn to make my own component, but since there is so few resources available I would love to see how yours is made.

**Reply by Kematia** [reply]
> Ciao @Kematia , your video preview looks very impressive. Hope you don't mind me asking, but just curious if you ever released the custom component? I'm trying to learn to make my own component, but since there is so few resources available I would love to see how yours is made.

In the end I went with the M2A builder (IIRC the m2a builder was not yet released when I was working on the project in the video). I should have it somewhere if you want to take a look at it (in case ping me on discord: `kematia#8232`). 

I went with the m2a because supporting the component was a bit too much work, and the custom extensions dev process was (and still is) a bit too slow for me.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
